tracing_test.go

```diff
@@ -21,314 +21,515 @@
+	return h.handleCall(ctx, req, func(ctx context.Context, req *TracingRequest) *TracingResponse {
+		jctx := ctx.(json.Context)
+		sc := h.ch.Peers().GetOrAdd(h.ch.PeerInfo().HostPort)
+		childResp := new(TracingResponse)
+		require.NoError(h.t, json.CallPeer(jctx, sc, h.ch.PeerInfo().ServiceName, "call", req, childResp))
+		return childResp
+	})
+}
 
-	var state struct {
-		signal chan struct{}
+func (h *JSONHandler) onError(ctx context.Context, err error) { h.t.Errorf("onError %v", err) }
 
-		call TraceData
-		span Span
-	}
-	testTraceReporter := TraceReporterFunc(func(data TraceData) {
-		defer close(state.signal)
+// ThriftHandler tests tracing over Thrift encoding
+type ThriftHandler struct {
+	traceHandler
+}
 
-		span := data.Span
-		data.Span = Span{}
-		state.call = data
-		state.span = span
+func (h *ThriftHandler) Call(ctx thrift.Context, arg *gen.Data) (*gen.Data, error) {
+	req := new(TracingRequest).fromThrift(arg)
+	res, err := h.handleCall(ctx, req, func(ctx context.Context, req *TracingRequest) *TracingResponse {
+		tctx := ctx.(thrift.Context)
+		res, err := h.thriftClient.Call(tctx, req.toThrift())
+		require.NoError(h.t, err)
+		return new(TracingResponse).fromThrift(h.t, res)
 	})
+	require.NoError(h.t, err)
+	return res.toThrift(h.t), nil
+}
 
-	traceReporterOpts := testutils.NewOpts().SetTraceReporter(testTraceReporter)
-	tests := []struct {
-		name       string
-		serverOpts *testutils.ChannelOpts
-		clientOpts *testutils.ChannelOpts
-		expected   []Annotation
-		fromServer bool
-	}{
-		{
-			name:       "inbound",
-			serverOpts: traceReporterOpts,
-			expected: []Annotation{
-				{Key: "sr", Timestamp: initialTime.Add(2 * time.Second)},
-				{Key: "ss", Timestamp: initialTime.Add(3 * time.Second)},
-			},
-			fromServer: true,
-		},
-		{
-			name:       "outbound",
-			clientOpts: traceReporterOpts,
-			expected: []Annotation{
-				{Key: "cs", Timestamp: initialTime.Add(time.Second)},
-				{Key: "cr", Timestamp: initialTime.Add(6 * time.Second)},
-			},
-		},
-	}
-
-	for _, tt := range tests {
-		serverNow, serverNowFn := testutils.NowStub(initialTime.Add(time.Second))
-		clientNow, clientNowFn := testutils.NowStub(initialTime)
-		serverNowFn(time.Second)
-		clientNowFn(time.Second)
+func (h *ThriftHandler) Simple(ctx thrift.Context) error {
+	return nil
+}
 
-		// Note: we disable the relay as the relay shares the same options
-		// and since the relay would call timeNow, it causes a mismatch in
-		// the expected timestamps.
-		tt.serverOpts = testutils.DefaultOpts(tt.serverOpts).SetTimeNow(serverNow).NoRelay()
-		tt.clientOpts = testutils.DefaultOpts(tt.clientOpts).SetTimeNow(clientNow)
+type TracerType string
 
-		WithVerifiedServer(t, tt.serverOpts, func(ch *Channel, hostPort string) {
-			state.signal = make(chan struct{})
+const (
+	Noop   TracerType = "NOOP"
+	Basic  TracerType = "BASIC"
+	Jaeger TracerType = "JAEGER"
```

Comment:
Adding comments. I chose strings because in case of test errors I get a nice log like this:
```
Messages:	Wrong span count; test {forwardCount:2 tracingDisabled:false format:raw expectedBaggage: expectedSpanCount:2} 
with tracer {tracerType:BASIC tracer:0xc820016a00 spansRecorded:0x16f870 resetSpans:0x16f8b0 isFake:false zipkinCompatible:false}
```
(note `tracerType:BASIC`), but if I use int/iota, then it prints ``tracerType:1` - ¯\_(ツ)_/¯